Enable change datatype in bulk

davelopez opened the pull request:
Closes #14033

This enables back the `Change datatype` bulk operation and tries to set the metadata for each dataset when doing so.

## TODO
- [x] Add API tests
- [x] Regenerate metadata in the *correct way*

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] Instructions for manual testing are as follows:
  

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

davelopez committed:
Enable `change_datatype` bulk operation

davelopez committed:
Validate `change_datatype` operation in API

Celery must be enabled and the data type must be registered

davelopez committed:
Show response error on bulk operation bad request

davelopez committed:
Set metadata on bulk change_datatype

davelopez commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1600,9 +1609,14 @@ def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
+            if params.datatype == "auto":  # Trigger job to auto-detect datatype
+                self.hda_manager.set_metadata(trans, item, overwrite=True)
+            else:
+                set_metadata.delay(self.hda_manager, None, dataset_id=item.id)
```
Is this the correct way of regenerating the metadata? It seems I'm not doing it right... cause I get `Object of type HDAManager is not JSON serializable` :thinking:

mvdbeek commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1600,9 +1609,14 @@ def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
+            if params.datatype == "auto":  # Trigger job to auto-detect datatype
+                self.hda_manager.set_metadata(trans, item, overwrite=True)
+            else:
+                set_metadata.delay(self.hda_manager, None, dataset_id=item.id)
```
```suggestion
                set_metadata.delay(dataset_id=item.id)
```
The rest is injected by lagom. I think the task itself may also need a flush. I haven't check carefully, but I haven't spotted where you are actually changing the datatype. The single HDA change datatype operation also checks that the dataset isn't currently used as an input or output dataset, I think we still need to do that.

davelopez replied:
> The rest is injected by lagom. 

Ohh right! I completely forgot about the injection :facepalm: Thank you!

> I think the task itself may also need a flush. 

Ok, I'll check if I can make sure the task makes the flush after finishing.

> I haven't check carefully, but I haven't spotted where you are actually changing the datatype.

I was doing it before, but I dropped it by mistake, thanks for noticing!

> The single HDA change datatype operation also checks that the dataset isn't currently used as an input or output dataset, I think we still need to do that.

Ok, I will try to that too :+1:

davelopez committed:
Update lib/galaxy/webapps/galaxy/services/history_contents.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

davelopez committed:
Add API test for bulk change datatype

davelopez committed:
Add API test to ensure error on use

davelopez committed:
Fixes on bulk change datatype

- Ensure the datatype can be changed
- Guess the type when `auto`
- Make the actual change through the `datatypes_registry`

davelopez committed:
Flush after set metada in celery task

davelopez committed:
Add API test for bulk auto-detect datatype

davelopez committed:
Add overwrite param to set_metadata in celery task

Include type hints

mvdbeek commented on lib/galaxy/celery/tasks.py:
```diff
@@ -103,15 +106,24 @@ def set_metadata(
+    model_class: str = "HistoryDatasetAssociation",
+    overwrite: bool = True,
 ):
+    dataset: Optional[model.DatasetInstance] = None
```
```suggestion
```

mvdbeek commented on lib/galaxy/celery/tasks.py:
```diff
@@ -103,15 +106,24 @@ def set_metadata(
+                    if spec.get("default"):
+                        setattr(dataset.metadata, name, spec.unwrap(spec.get("default")))
+        dataset.datatype.set_meta(dataset)
+        sa_session.flush()
```
```suggestion
    else:
        raise NotImplementedError(f"Cannot set metadata for model_class {model_class}")
    if overwrite:
        for name, spec in dataset.metadata.spec.items():
            # We need to be careful about the attributes we are resetting
            if name not in ["name", "info", "dbkey", "base_name"]:
                if spec.get("default"):
                    setattr(dataset.metadata, name, spec.unwrap(spec.get("default")))
    dataset.datatype.set_meta(dataset)
    sa_session.flush()
```

mvdbeek commented on lib/galaxy/celery/tasks.py:
```diff
@@ -2,7 +2,10 @@
+from typing import (
+    Callable,
+    Optional,
+)
```
```suggestion
from typing import Callable
```

davelopez committed:
Apply suggestions from code review

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

davelopez committed:
Add set_peek to set_metadata celery task

mvdbeek commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1600,9 +1610,30 @@ def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
+        else:
+            datatype = datatype
+        trans.app.datatypes_registry.change_datatype(hda, datatype)
+        trans.sa_session.flush()
```
Can you create a change_datatype task and push that there ? The sniffing can cause significant I/O.

davelopez replied:
Sure! I can try

davelopez replied:
Done! I learned a bit of celery today, thank you! :smile:

mvdbeek commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1600,9 +1610,30 @@ def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
         if isinstance(item, HistoryDatasetAssociation):
-            item.change_datatype(params.datatype)
+            self._ensure_can_change_hda_datatype(item)
+            self._set_hda_datatype(item, params.datatype, trans)
```
```suggestion
            item.dataset.state = item.dataset.states.SETTING_METADATA
            self._set_hda_datatype(item, params.datatype, trans)
```
For large files this can still take a while, so we should switch to the SETTING_METADATA state here, which means the dataset is considered "pending".

mvdbeek commented on lib/galaxy/celery/tasks.py:
```diff
@@ -100,13 +100,29 @@ def set_job_metadata(
+                    setattr(dataset.metadata, name, spec.unwrap(spec.get("default")))
     dataset.datatype.set_meta(dataset)
+    dataset.set_peek()
+    sa_session.flush()
```
```suggestion
    try:
        dataset.datatype.set_meta(dataset)
        dataset.set_peek()
        dataset.dataset.state = dataset.dataset.states.OK
    except Exception:
        log.info(f"Setting metadata failed on {model_class} {dataset.id}: {str(e)}")
        dataset.dataset.state = dataset.dataset.states.FAILED_METADATA
    sa_session.flush()
```

davelopez committed:
Apply suggestions from code review

Add set_meta exception handling in celery task and mark dataset state as SETTING_METADATA for large files.

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

davelopez committed:
Refactor extract _get_dataset_by_id method

davelopez committed:
Add celery task to change datatype

davelopez committed:
Chain change_datatype and set_metadata tasks

Use immutable signature to avoid passing the result of the first task to the next because this will disrupt the dependency injection.

mvdbeek commented on lib/galaxy/celery/tasks.py:
```diff
@@ -98,15 +99,61 @@ def set_job_metadata(
+    if datatype == "auto":
+        path = dataset_instance.dataset.file_name
+        datatype = sniff.guess_ext(path, datatypes_registry.sniff_order)
+    datatypes_registry.change_datatype(dataset_instance, datatype)
+    sa_session.flush()
```
I'd call directly into the set_metadata task here. If you call it without `delay` it runs like your normal function.

davelopez replied:
Ohh, was just trying to chain both tasks but took me some time until I figured out the chaining was messing with the dependency injection and need to use `immutable signatures` :sweat_smile: 

I think it works now, do you still think is better to call it directly in set_metadata?

mvdbeek replied:
Yes, those don't need to be separate tasks, they have significant overhead and it's really one atomic operation.

davelopez replied:
Makes total sense, done!

davelopez committed:
Reduce duplication a bit

mvdbeek commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1602,7 +1612,13 @@ def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
-            item.change_datatype(params.datatype)
+            self.hda_manager.ensure_can_change_datatype(item)
+            self.hda_manager.ensure_can_set_metadata(item)
+            item.dataset.state = item.dataset.states.SETTING_METADATA
```
You need to flush when you make a change.

davelopez replied:
There is a `global` flush at the end of the request here:

https://github.com/davelopez/galaxy/blob/09543032c7b91bc4ce6b3b488974d8cf3df114fc/lib/galaxy/webapps/galaxy/services/history_contents.py#L710

Should I flush for every item instead in this case?

davelopez replied:
I added the flush here too, just in case, I can revert if the global request flush is enough.

mvdbeek replied:
Yeah, I think we need it here since the task can in principle (not very likely ... but still possible) start, or worse finish, before the request finishes.

mvdbeek commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1602,7 +1612,13 @@ def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
+            # Chain tasks using immutable signature to avoid messing with dependency injection
+            change_datatype_task = change_datatype.si(dataset_id=item.id, datatype=params.datatype)
+            set_metadata_task = set_metadata.si(dataset_id=item.id)
+            (change_datatype_task | set_metadata_task).delay()
```
There's a cost to having separate tasks, I think `change_datatype` should just do it all in one task.

davelopez committed:
Set metadata directly in change datatype task

davelopez committed:
Flush dataset state change before running task
